tensorboard/data/server/commit.rs

```diff
@@ -0,0 +1,138 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+//! Shared state for sampled data available to readers.
+
+use std::collections::HashMap;
+use std::sync::RwLock;
+
+use crate::proto::tensorboard as pb;
+use crate::reservoir::Basin;
+use crate::types::{Run, Step, Tag, WallTime};
+
+/// Current state of in-memory sampled data.
+///
+/// A commit is an internally mutable structure. All readers and writers should keep a shared
+/// reference to a single commit. When writers need to update it, they grab an exclusive lock to
+/// the contents.
+///
+/// Deadlock safety: any thread should obtain the outer lock (around the hash map) before an inner
+/// lock (around the run data), and should obtain at most one `RunData` lock at once.
+#[derive(Debug, Default)]
+pub struct Commit {
+    pub runs: RwLock<HashMap<Run, RwLock<RunData>>>,
+}
+
+impl Commit {
+    /// Creates a new, empty commit.
+    pub fn new() -> Self {
+        Commit::default()
+    }
+}
+
+/// Data for a single run.
+///
+/// This contains all data and metadata for a run. For now, that data includes only scalars;
+/// tensors and blob sequences will come soon.
+#[derive(Debug, Default)]
+pub struct RunData {
+    /// The time of the first event recorded for this run.
+    ///
+    /// Used to define an ordering on runs that is stable as new runs are added, so that existing
+    /// runs aren't constantly changing color.
+    pub start_time: Option<WallTime>,
+
+    /// Scalar time series for this run.
+    pub scalars: TagStore<ScalarValue>,
+}
+
+pub type TagStore<V> = HashMap<Tag, TimeSeries<V>>;
+
+#[derive(Debug)]
+pub struct TimeSeries<V> {
+    /// Summary metadata for this time series.
+    pub metadata: Box<pb::SummaryMetadata>,
```

Comment:
You’re right that it’s not that bad. Here are the ones that I’m
concerned about, with output from `rustc -Zprint-type-sizes`:

<details>

<summary>Type size information</summary>

```
print-type-size type: `proto::tensorboard::summary::Value`: 536 bytes, alignment: 8 bytes
print-type-size     field `.node_name`: 24 bytes
print-type-size     field `.tag`: 24 bytes
print-type-size     field `.metadata`: 104 bytes
print-type-size     field `.value`: 384 bytes
```

```
print-type-size type: `proto::tensorboard::TensorProto`: 376 bytes, alignment: 8 bytes
print-type-size     field `.tensor_shape`: 32 bytes
print-type-size     field `.tensor_content`: 24 bytes
print-type-size     field `.half_val`: 24 bytes
print-type-size     field `.float_val`: 24 bytes
print-type-size     field `.double_val`: 24 bytes
print-type-size     field `.int_val`: 24 bytes
print-type-size     field `.string_val`: 24 bytes
print-type-size     field `.scomplex_val`: 24 bytes
print-type-size     field `.int64_val`: 24 bytes
print-type-size     field `.bool_val`: 24 bytes
print-type-size     field `.dcomplex_val`: 24 bytes
print-type-size     field `.resource_handle_val`: 24 bytes
print-type-size     field `.variant_val`: 24 bytes
print-type-size     field `.uint32_val`: 24 bytes
print-type-size     field `.uint64_val`: 24 bytes
print-type-size     field `.dtype`: 4 bytes
print-type-size     field `.version_number`: 4 bytes
```

```
print-type-size type: `proto::tensorboard::SummaryMetadata`: 104 bytes, alignment: 8 bytes
print-type-size     field `.plugin_data`: 48 bytes
print-type-size     field `.display_name`: 24 bytes
print-type-size     field `.summary_description`: 24 bytes
print-type-size     field `.data_class`: 4 bytes
print-type-size     end padding: 4 bytes
```

</details>

So yes, “large amount of data” can definitely go higher than 104 bytes,
but that’s the only reason here. It should also make hash map resizes of
the `TagStore` a bit faster, though that shouldn’t matter too much.

Boxing all three of these together improved end-to-end performance by
~25%, and I would believe that most of that is due to `TensorProto` and
`Value` since those directly affect the reservoir size, but I also don’t
mind doing this. If we want to unbox this, that’s okay with me.

Reply:
Okay, it was primarily a curiosity question - fine by me to leave it boxed.  I didn't realize it'd be 24 bytes per Vec aka per String so it does add up to more than I'd have thought; at least one reason is that in my head I'm still thinking about immutable string types where there isn't an extra 8 bytes for capacity.

I doubt that it makes a big difference in our case, but in general it would seem like some applications with a lot of  immutable strings would save space overhead by having an immutable owned string representation with pointer + length only?  I guess that must just not come up very often since I don't see much evidence that this exists.

Reply:
And, of course, not all languages with immutable strings fit them into 8
or 16 bytes:

```python
>>> import sys
>>> sys.getsizeof("")
49
```

Reply:
Interesting, I think `Box<str>` is what I was looking for, and now that you mention it I see there's even https://doc.rust-lang.org/std/string/struct.String.html#method.into_boxed_str.

It hadn't occurred to me since `str` on its own seems a little exotic, but this explanation helps.  What didn't help was the fact that https://doc.rust-lang.org/book/ch04-03-slices.html starts by saying "Another data type that does not have ownership is the _slice_" when I think what they're really doing is referring to `&[_]` as a "slice" informally and eliding the distinction between that and `[_]`.  So whether or not a "slice" in the strict meaning of `[_]` has ownership depends on the type of pointer being used to hold it (namely if it's `&`, `&mut`, or `Box`).